data/src/main/scala/ch.epfl.scala.index.data/maven/PomsReader.scala

```diff
@@ -27,27 +29,30 @@ object PomsReader {
 
   def loadAll(
       paths: DataPaths
-  ): List[Try[(ReleaseModel, LocalRepository, String)]] = {
+  ): Iterable[(ReleaseModel, LocalRepository, String)] = {
+    import ExecutionContext.Implicits._
     import LocalPomRepository._
 
-    val centralPoms = PomsReader(MavenCentral, paths).load()
-    val centralShas = centralPoms.collect { case Success((_, _, sha)) => sha }.toSet
-
-    val bintrayPoms = PomsReader(Bintray, paths).load().filter {
-      case Success((_, _, sha)) => !centralShas.contains(sha)
-      case _                    => true
-    }
-    val bintrayShas = bintrayPoms.collect { case Success((_, _, sha)) => sha }.toSet
-
-    val usersPoms = PomsReader(UserProvided, paths).load().filter {
-      case Success((_, _, sha)) =>
-        !centralShas.contains(sha) && !bintrayShas.contains(sha)
-      case _ => true
+    val centralPomsF = PomsReader(MavenCentral, paths).iterator()
+    val bintrayPomsF = PomsReader(Bintray, paths).iterator()
+    val usersPomsF = PomsReader(UserProvided, paths).iterator()
+    val ivysDescriptors = SbtPluginsData(paths.ivysData).iterator
+
+    val allPoms = for {
+      centralPoms <- centralPomsF
+      bintrayPoms <- bintrayPomsF
+      usersPoms <- usersPomsF
+    } yield {
+      (centralPoms ++ bintrayPoms ++ usersPoms ++ ivysDescriptors)
+        .foldLeft(Map[String, (ReleaseModel, LocalRepository, String)]()) {
+          case (acc, pom @ (_, _, sha)) =>
+            if (acc.contains(sha)) acc
+            else acc + (sha -> pom)
+        }
+        .values
     }
 
-    val ivysDescriptors = SbtPluginsData(paths.ivysData).load()
-
-    centralPoms ::: bintrayPoms ::: usersPoms ::: ivysDescriptors
+    Await.result(allPoms, Duration.Inf)
```

Comment:
OPt Why doing the await here and not higher?

Reply:
I wanted to maintain the method signature. But now that I read the code again, I think that `ReadPoms.iterator` being a future is nonsense. I will change that and move the Await out of it.